src/VisualStudio/Core/Def/Implementation/ChangeSignature/ChangeSignatureDialogViewModel.ParameterViewModels.cs

```diff
@@ -0,0 +1,246 @@
+                    {
+                        return true;
+                    }
+
+                    if (this == ChangeSignatureDialogViewModel._parametersWithoutDefaultValues.LastOrDefault() &&
+                        (ChangeSignatureDialogViewModel._parametersWithDefaultValues.Any() || ChangeSignatureDialogViewModel._paramsParameter != null))
+                    {
+                        return true;
+                    }
+
+                    if (this == ChangeSignatureDialogViewModel._parametersWithDefaultValues.LastOrDefault() &&
+                        ChangeSignatureDialogViewModel._paramsParameter != null)
+                    {
+                        return true;
+                    }
+
+                    return false;
+                }
+            }
+
+            public virtual string CallSiteValue => string.Empty;
+        }
+
+        public class AddedParameterViewModel : ParameterViewModel
+        {
+            public override Parameter Parameter => _addedParameter;
+            private readonly AddedParameter _addedParameter;
+
+            public AddedParameterViewModel(ChangeSignatureDialogViewModel changeSignatureDialogViewModel, AddedParameter addedParameter)
+                : base(changeSignatureDialogViewModel)
+            {
+                _addedParameter = addedParameter;
+            }
+
+            /// <summary>
+            /// Display text for the type in the UI, including a marker if it doesn't bind.
+            /// </summary>
+            public override string Type => _addedParameter.TypeName;
+
+            public override Visibility TypeWarningVisibility => _addedParameter.TypeBinds ? Visibility.Collapsed : Visibility.Visible;
+
+            public override string ParameterName => _addedParameter.Name;
+
+            public override bool IsRemoved { get => false; set => throw new InvalidOperationException(); }
+
+            public override string FullAutomationText
+            {
+                get
+                {
+                    var baseText = base.FullAutomationText;
+                    return ServicesVSResources.Added_Parameter + baseText + string.Format(ServicesVSResources.Inserting_call_site_value_0, CallSite);
+                }
+            }
+
+            public override bool IsDisabled => false;
+
+            public override string CallSite
+            {
+                get
+                {
+                    return _addedParameter.UseNamedArguments
+                        ? _addedParameter.Name + ": " + _addedParameter.CallSiteValue
+                        : _addedParameter.CallSiteValue;
+                }
+            }
+
+            public override string CallSiteValue => _addedParameter.CallSiteValue;
+
+            public override string InitialIndex => "+";
+
+            // Newly added parameters cannot have modifiers yet
+            public override string Modifier => string.Empty;
+
+            // Only required parameters are supported currently
```

Comment:
not sure I understand this comment. Wouldn't that mean that the `IsRequired` below should always return true?

Reply:
Stale comment deleted.